prebid-server/pbs-modules/rules-engine.md

```diff
@@ -0,0 +1,805 @@
+         1. userFpdAvailable=wildcard  
+            1. remove bidderA (leaf 1\)  
+      2. eidAvailable(sharedId)=false  
+         1. userFpdAvailable=true  
+            1. remove bidderA, bidderD (leaf 2\)  
+         2. userFpdAvailable=false  
+            1. remove bidderA, bidderB  (leaf 3\)  
+   2. channel=wildcard  
+      1. eidAvailable(sharedId)=wildcard  
+         1. userFpdAvailable=wildcard  
+            1. remove bidderC (leaf 4\)  
+2. countryGroup=APAC (EU countries false, APAC countries true)  
+   1. channel=web  
+      1. eidAvailable(sharedId)=true  
+         1. userFpdAvailable=true  
+            1. remove bidderA, bidderE (leaf 5\)  
+         2. userFpdAvailable=false  
+            1. remove bidderA, bidderB, bidderE  (leaf 6\)  
+      2. eidAvailable(sharedId)=false  
+         1. userFpdAvailable=wildcard  
+            1. remove bidderA, bidderD, bidderF  (leaf 7\)  
+   2. channel=wildcard  
+      1. eidAvailable(sharedId)=wildcard  
+         1. userFpdAvailable=true  
+            1. remove bidderA, bidderC  (leaf 8\)  
+         2. userFpdAvailable=false  
+            1. remove bidderA, bidderF  (leaf 9\)  
+3. countryGroup=OTHER (EU countries false, APAC countries false)  
+   1. channel=web  
+      1. eidAvailable(sharedId)=wildcard  
+         1. userFpdAvailable=wildcard  
+            1. no removals  (leaf 10\)  
+   2. channel=wildcard  
+      1. eidAvailable(sharedId)=true  
+         1. userFpdAvailable=wildcard  
+            1. remove bidderB, bidderE (leaf 11\)  
+      2. eidAvailable(sharedId)=false  
+         1. userFpdAvailable=wildcard  
+            1. remove bidderE, bidderF (leaf 12\)
+
+Rules engine config:
+
+```json5
+{
+  "hooks": {
+    "modules": {
+      "pb-rules-engine": {
+        "enabled": true,
+        "timestamp": "20250314 00:00:00",
+        "ruleSets": [
+          {
+            "stage": "processed-auction-request",
+            "name": "remove-bidder-by-country-channel-eid-userFpd",
+            "version": "5678",
+            "modelGroups": [
+              {
+                "weight": 100,
+                "analyticsKey": "filter-bidder",
+                "version": "4567",
+                "schema": [{"function": "deviceCountry", "args": [["FRA","DEU","GBR"]]},
+                           {"function": "deviceCountry", "args": [["JPN","AUS","KOR"]]},
+                           {"function": "channel"},
+                           {"function": "eidAvailable", "args": [["pubcid.org"]]},
+                           {"function": "userFpdAvailable" }
+                ],
+                "default": [],
+                "rules": [{
+                    "conditions": ["true","false","app","true","*"],
+                    "results": [{
+                        "function": "excludeBidders",
+                        "args": [{
+                            "bidders": [ "bidderA"],
+                            "analyticsValue": "leaf1"
+                        }]
```

Comment:
`"args"` is an object. Remove square brackets for this leaf and all other 11 leaves in this example.